components/service-operator/internal/aws/cloudformation_controller.go

```diff
@@ -0,0 +1,191 @@
+/*
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package aws
+
+import (
+	"context"
+	"fmt"
+	"strings"
+
+	"github.com/alphagov/gsp/components/service-operator/internal"
+	"github.com/go-logr/logr"
+	ctrl "sigs.k8s.io/controller-runtime"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/session"
+	awscloudformation "github.com/aws/aws-sdk-go/service/cloudformation"
+	goformation "github.com/awslabs/goformation/cloudformation"
+	"github.com/awslabs/goformation/cloudformation/resources"
+)
+
+type CloudFormationTemplate interface {
+	Template(string, []resources.Tag) *goformation.Template
+	Parameters() ([]*awscloudformation.Parameter, error)
+	ResourceType() string
+}
+
+type CloudFormationReconciler interface {
+	Reconcile(context.Context, logr.Logger, ctrl.Request, CloudFormationTemplate, bool) (internal.Action, StackData, error)
+}
+
+type CloudFormationController struct {
+	ClusterName string
+}
+
+var (
+	nonUpdatable = []string{
+		awscloudformation.StackStatusCreateInProgress,
+		awscloudformation.StackStatusRollbackInProgress,
+		awscloudformation.StackStatusDeleteInProgress,
+		awscloudformation.StackStatusUpdateInProgress,
+		awscloudformation.StackStatusUpdateCompleteCleanupInProgress,
+		awscloudformation.StackStatusUpdateRollbackInProgress,
+		awscloudformation.StackStatusUpdateRollbackCompleteCleanupInProgress,
+		awscloudformation.StackStatusReviewInProgress,
+		awscloudformation.StackStatusDeleteComplete,
+	}
+)
+
+func (r *CloudFormationController) Reconcile(ctx context.Context, log logr.Logger, req ctrl.Request, cloudFormationTemplate CloudFormationTemplate, deleting bool) (internal.Action, StackData, error) {
+	stackName := fmt.Sprintf("%s-%s-%s-%s-%s", r.ClusterName, "gsp-service-operator", cloudFormationTemplate.ResourceType(), req.Namespace, req.Name)
```

Comment:
Very nice!

But I'm not sure the "gsp-service-operator" token is needed here. :man_shrugging:

Reply:
Agree, the resources are tagged with some values, which may make it obsolete from the stackName :)